content/blog/2021-11-10-agent.md

```diff
@@ -0,0 +1,156 @@
+
+![Yoda](/assets/blog/2021-11-10/yoda.gif)
+
+The other interesting use case that emerges is the notion of **Edge** clusters or networks. With the industries like telecommunication, automotive and IoT devices adopting cloud-native technologies, we see more and more much smaller clusters with a restricted amount of resources. This is forcing all data (including observability) to be transferred to remote, bigger counterparts as almost nothing can be stored on those remote nodes.
+
+What does that mean? That means monitoring data has to be somehow aggregated, presented to users and sometimes even stored on the *global* level. This is often called a **Global-View** feature.
+
+Naively, we could think about implementing this by either putting Prometheus on that global level and scraping metrics across remote networks or pushing metrics directly from the application to the central location for monitoring purposes. Let me explain why both are generally *very* bad ideas:
+
+* Scrape across the network is a terrible idea as we employ a huge amount of unknowns in our monitoring pipeline. The local pull model allows Prometheus to know why exactly the metric target has problems and when. Maybe it's down, misconfigured, restarted, too slow to give us metrics (e.g. CPU saturated), not discoverable by service discovery, we don't have credentials to access or just DNS, network, or whole cluster is down. By putting our scraper outside of the network, we risk losing some of the information due to unreliable scrape and irregular scrape intervals. On top of that, we risk losing important visibility completely if the network is temporarily down. Please don't do it. It's not worth it. (:
+* Metric push directly from the application to some central location is equally bad. Especially when you monitor a larger fleet, you know nothing literally when you don't see metrics from remote applications. Is the application down? Is my receive pipeline down? Maybe the application failed to authorize? Maybe it failed to get the IP address of my remote cluster? Maybe it's too slow? Maybe the network is down? Such design is often a recipe for a failure (or at least shrug 🤷🏽‍♀️ moments).
+
+Prometheus introduced three main protocols to support the global view case, as discussed many times at various conferences. Each with its own pros and cons. Let's go briefly go through those. They are presented in orange colour on the diagram below.
+
+![Prometheus global view](/assets/blog/2021-11-10/prom-remote.png)
+
+* **Federation** was introduced as the first feature for aggregation purposes. It allows a special scrape done by Prometheus on a global level for a subset of metrics in the leaf Prometheus. Such a "federation" scrape reduces some unknowns across networks because metrics exposed by federate endpoints inject timestamps. Yet, it usually suffers from the inability to federate all metrics and not lose data during longer network partitions (minutes).
```

Comment:
```suggestion
* **Federation** was introduced as the first feature for aggregation purposes. It allows a global-level Prometheus server to scrape a subset of metrics from a leaf Prometheus. Such a "federation" scrape reduces some unknowns across networks because metrics exposed by federation endpoints inject include the original samples' timestamps. Yet, it usually suffers from the inability to federate all metrics and not lose data during longer network partitions (minutes).
```